lib/cchain/provider/abci.go

```diff
@@ -138,15 +139,119 @@ func newABCIFetchFunc(cl atypes.QueryClient) fetchFunc {
 		})
 		if err != nil {
 			incQueryErr(endpoint)
-			return nil, errors.Wrap(err, "abci query approved-from")
+			return nil, errors.Wrap(err, "abci query attestations-from")
 		}
 
-		atts, err := atypes.AttestationsFromProto(resp.Attestations)
+		if len(resp.Attestations) != 0 {
+			atts, err := atypes.AttestationsFromProto(resp.Attestations)
+			if err != nil {
+				return nil, errors.Wrap(err, "attestations from proto")
+			}
+
+			return atts, nil
+		}
+
+		earliestAttestationAtLatestHeight, ok, err := queryEarliestAttestation(ctx, cl, chainVer, 0)
 		if err != nil {
-			return nil, errors.Wrap(err, "attestations from proto")
+			incQueryErr(endpoint)
+			return nil, errors.Wrap(err, "abci query earliest-attestation-in-state")
+		}
+
+		// Either no attestations have happened yet, or the queried fromOffset is in the "future"
+		// Caller has to wait and retry in both cases
+		if !ok || earliestAttestationAtLatestHeight.BlockHeader.BlockOffset < fromOffset {
+			// First attestation hasn't happened yet, return empty
+			return []xchain.Attestation{}, nil
+		}
+
+		// Exponentially backoff to find a good start point for binary search, this prefers more recent queries
+		info, err := client.ABCIInfo(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "abci query info")
+		}
+		startHeightIndex := 0
+		endHeightIndex := int(info.Response.LastBlockHeight)
+		lookback := 1
+		lookbackStepsCounter := 0 // For metrics only
+		queryHeight := endHeightIndex - lookback
+		for queryHeight > 0 {
+			lookbackStepsCounter++
+			earliestAtt, ok, err := queryEarliestAttestation(ctx, cl, chainVer, queryHeight)
+			if err != nil {
+				incQueryErr(endpoint)
+				return nil, errors.Wrap(err, "abci query earliest-attestation-in-state")
+			}
+
+			// If we're before the first attestation, or found an earlier attestation, it's a good start height
+			if !ok || earliestAtt.BlockHeader.BlockOffset <= fromOffset {
+				startHeightIndex = queryHeight
+				break
+			}
+
+			// Otherwise, keep moving back
+			endHeightIndex = queryHeight
+			lookback *= 2
+			queryHeight -= lookback
 		}
 
-		return atts, nil
+		log.Debug(ctx, "Cchain fetch lookback done ", "steps", lookbackStepsCounter)
```

Comment:
lets add metrics for the searching steps: lookback_steps and search_steps histogram.